### DISCUSSION TITLE
How to enforce different cache keys?

### DISCUSSION THREAD

**Discussion by zoontek** [urql-graphql/urql#1738] [Q&A] [answered]
Hi 👋

We have a complex project with multiple GQL endpoints. Our product is a GQL API, on our dashboard we need to connect to the `/live` one, to the `sandbox` and to the `/admin`. Because it's hard to have multiple GQL clients, we stick with one and play with context instead, but it cause cache issues because the operation key does not change if the endpoint / headers are differents (as seen here https://github.com/FormidableLabs/urql/blob/main/packages/core/src/utils/request.ts).

For now we use an ugly hack and add the `userId` and `url` to our GQL document:

```ts
const getContext = (userId): OperationContext => ({
  requestPolicy: "cache-first",
  url: env.SANDBOX_URL,
  fetchOptions: { headers: { "impersonated-user-id": userId } },
});

// …

const [{ fetching, data, error }] = useExampleQuery({
  context: useMemo(() => getContext(userId), [userId]),
  variables: {
    // needed to avoid cache clashes
    // @ts-ignore
    userId,
    // @ts-ignore
    url: env.SANDBOX_URL,

    realInputData,
  },
});
```

We also though about writing a custom exchange to replace the key before everything else, but it seems dangerous.
Does a proper way to derive the key from document + url + some headers exists?

Thanks!

**Comment by kitten** [comment, marked as answer]
Well, I hope, I don't have to mention my usual spiel about how it's important to minimise and unify GraphQL APIs to reflect your actual org chart into one, to not lose any of the benefits of GraphQL, but I suppose the decision's been made. I'd usually then say to actually segregate your app into different clients, but that boat has left the port too 😅 

So, this isn't a recommendation of best practices, and I wouldn't recommend this otherwise, but it is possible to monkey patch `client.createRequestOperation` to achieve this: https://github.com/FormidableLabs/urql/blob/ec61177828943fcaf18368311ec236cf4e6666d3/packages/core/src/client.ts#L293-L295

This method combines a request with the `Client`'s base `context` and the passed in partial `context` to create an operation. What this means is that if you wrap this function you could alter `Operation#key` to be namespaced by your `context.url`. Mind you that the `key` is a hash and hence a number, but you can add additional information to a hash like so: https://github.com/FormidableLabs/urql/blob/ec61177828943fcaf18368311ec236cf4e6666d3/packages/core/src/utils/hash.ts#L4-L11

Again, I'm just offering you a shortterm solution here, since that's what you likely want ❤️ But this doesn't really sound sustainable

Edit: I should probably clarify, we actually worked around monkey patching, so this isn't as easy as I thought it'd be without copying the client source code entirely. Let me think about this again

**Reply by zoontek** [reply]
Thanks for the tip, monkey patching should work!

For the details, `/live` and `/sandbox` expose the same schema, but they return different data (fake ones to facilitate the API integration or live data - like Stripe does). The `/admin` is used for specific mutations that cannot be done outside of our dashboard. We didn't really see how this could be done with a single enpoint.

As urql use React Context (with `Provider`), I didn't think multiple clients was doable. But indeed, it's too late 😅

**Reply by kitten** [reply]
So, I suppose the main problem for you right now is that `/live` and `/sandbox` are used potentially at the same time and with the same inputs? is that correct? I don't really see a danger with `/admin` since it'll have a unique schema

**Reply by zoontek** [reply]
Yes, that's our main issue. `/admin` could be at risk too when you impersonate a user using a HTTP header.

**Reply by zoontek** [reply]
@kitten Would you accept a MR which also uses the url to hash the key (I think it's a proper issue to have these clash when using context)?
I could also add a `ClientOption` to include headers in the computation too, in order to solve the "impersonate the user" issue.

```ts
export const urql = new Client({
  exchanges: [dedupExchange, cacheExchange, fetchExchange],
  url: env.API_URL,
  keyHeaders: ["foo", "bar"], // not sure about the name at all
});
```

**Reply by zoontek** [reply]
I would also be easier to avoid cache clashes on logout & re-login with another account, by just setting, for example with Hasura `keyHeaders: ["X-Hasura-Admin-Secret"]`

**Reply by kitten** [reply]
@zoontek Unfortunately not, no 😅 This assumption may hold in some cases, like for the default cache exchange, but it'd slowly break down, at the very latest when you reach Graphcache, so given that it'd be a workaround in a territory that we see as "non-idiomatic GraphQL" we're very hesitant to add any changes that cater for multiple identical schemas being treated differently per URL. The reason for that is simply that the schema doesn't represent your data anymore, and instead, you're providing inputs as that header and the URL itself. This also presents a challenge that it'd be possible to alter keys by infinite context options (that can also be nested, like `fetchOptions`, which may even be a function) in the `Client` itself.

So I see this as incompatible with our API from multiple angles. There are alternatives that I'd propose you could explore, but most would probably be based on an advice to add a special directive that a custom exchange could handle (which also solved the keying issue)
